### PR TITLE
Added the logic to increase the response cache maxAge for /schedule/id

### LINKED ISSUE #10321: Increase response cache max-age for `/schedules/{id}`
### Problem

`/schedules/{id}` is in among the most queried endpoints, the response cache header is set with the default 1 second max-age. This can cause a lot of unnecessary queries sent to the backend when the schedule has already finalized (either executed or expired).

In order to reduce the load, a much longer max-age (e.g., 1 hour) should be used instead.

### Solution

Set a longer max-age for `/schedules/{id}` response if any of the following is true

- `deleted` is true
- `executed_timestamp` is not null
- `executed_timestamp` is null, and one of the following is true
  - `expiration_time` is null and the wall clock time is at least 31 minutes after `consensus_timestamp`
  - `expiration_time` is not null and the wall clock time is at least 1 minute after `expiration_time`

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by HarshSawarkar**
**Description**:

Increase cache max-age to 1 hour for finalized schedules to reduce unnecessary backend queries.

**Related issue(s)**:

Fixes #10321 

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-mirror-rest/schedules.js [hunk 1]
```diff
@@ -134,7 +134,27 @@ const getScheduleById = async (req, res) => {
...
+  const schedule = rows[0];
+  res.locals[constants.responseDataLabel] = formatScheduleRow(schedule);
+
+  let maxAge = 1;
+  const now = Date.now() / 1000;
+  const executedTimestamp = utils.nsToSecNs(schedule.executed_timestamp);
+  const expirationTime = utils.nsToSecNs(schedule.expiration_time);
+  const consensusTimestamp = utils.nsToSecNs(schedule.consensus_timestamp);
+
+  if (schedule.deleted || (executedTimestamp && executedTimestamp !== '')) {
```
Comment (jnels124) [inline hedera-mirror-rest/schedules.js:146]: `executedTimestamp` is sufficient. `''` is falsey

Comment (jnels124) [reply hedera-mirror-rest/schedules.js:146]: This is a little easier to understand.

```bash
const hasExecuted = !!executedTimestamp;
const hasAutoExpired = !schedule.expiration_time && now >= consensusTimestamp + 1860;
const hasExpired = expirationTime && now >= expirationTime + 60;
const maxAge = hasExecuted || hasAutoExpired || hasExpired ? 3600 : 1;
```

File: hedera-mirror-rest/__tests__/specs/schedules/{id}/responseHeaders.json
Comment (jnels124) [inline hedera-mirror-rest/__tests__/specs/schedules/{id}/responseHeaders.json:1]: I believe this will break other tests as not all tests should result in the longer max age. Tests that rely on the different header value will need to be grouped into the same folder with each folder having the correct `responseHeaders.json`.

File: hedera-mirror-rest/__tests__/specs/schedules/{id}/executed.json
Comment (jnels124) [inline hedera-mirror-rest/__tests__/specs/schedules/{id}/executed.json:1]: We should add a test case for explicitly expired and one for implicitly expired.

Comment (jnels124) [reply hedera-mirror-rest/__tests__/specs/schedules/{id}/executed.json:1]: can look in the specs for examples of.

```
"features": {
      "fakeTime": "TIME_VALUE"
 },
```

**Review by jnels124 — CHANGES_REQUESTED** [review]

File: hedera-mirror-rest/schedules.js [hunk 2]
```diff
@@ -134,7 +134,21 @@ const getScheduleById = async (req, res) => {
...
+
+  const now = Date.now() / 1000;
+  const executedTimestamp = utils.nsToSecNs(schedule.executed_timestamp);
+  const expirationTime = utils.nsToSecNs(schedule.expiration_time);
+  const consensusTimestamp = utils.nsToSecNs(schedule.consensus_timestamp);
+
+  const hasExecuted = !!executedTimestamp || schedule.deleted;
+  const hasAutoExpired = !expirationTime && now >= consensusTimestamp + 1860;
+  const hasExpired = expirationTime && now >= expirationTime + 60;
+  const maxAge = hasExecuted || hasAutoExpired || hasExpired ? 3600 : 1;
```
Comment (xin-hedera) [inline hedera-mirror-rest/schedules.js:148]: - pull the logic to its own function so we can unit test all possible cases
- not a fan of converting a  timestamp of type `BigInt` to a string in the format of `ssss.nnnn`, and in fact there is bug in `consensusTimestamp + 1860` since it'll end up appending string `1860` to `consensusTimestamp`. Instead, use `utils.nowInNs()`, and use the nanosecond timestamp values in `schedule`, and finally add 31 minutes in nanosecond / 60 seconds in nanosecond for the relevant arithmetic operations
- we should not explicitly set `maxAge` if the conditions do not warrant a 3600 secs max-age. logic in responseHander will then set a default max-age from configuration file. Check the logic in transactions.js, `getTransactionsByIdOrHash`, around when `getTransactionsByIdOrHashCacheControlHeader` is called

File: hedera-mirror-rest/schedules.js [hunk 3]
```diff
@@ -134,7 +134,21 @@ const getScheduleById = async (req, res) => {
     throw new NotFoundError();
   }
 
+  const schedule = rows[0];
+  res.locals[constants.responseDataLabel] = formatScheduleRow(schedule);
```
Comment (xin-hedera) [inline hedera-mirror-rest/schedules.js:138]: this is duplicate of line 150

**Review by xin-hedera — CHANGES_REQUESTED** [review]

File: hedera-mirror-rest/schedules.js (hunk 2 above)
Comment (HarshSawarkar) [reply hedera-mirror-rest/schedules.js]: Updated. I initially overlooked that adding an integer to a string would result in concatenation. Are there any other edge cases I should cover in unit tests?

File: hedera-mirror-rest/schedules.js [hunk 4]
```diff
@@ -134,7 +134,29 @@ const getScheduleById = async (req, res) => {
     throw new NotFoundError();
   }
 
-  res.locals[constants.responseDataLabel] = formatScheduleRow(rows[0]);
+  const schedule = rows[0];
+  const maxAge = calculateScheduleMaxAge(schedule);
+  if (maxAge !== undefined) {
+    res.set('Cache-Control', `public, max-age=${maxAge}`);
```
Comment (xin-hedera) [inline hedera-mirror-rest/schedules.js:140]: cache-control header set here most likely will get overridden by the default value, in the logic in `responseHandler` since it runs at the end.

Instead, should do it the same way as in `transactions.js`, i.e.

```javascript
  res.locals[constants.responseHeadersLabel] = getTransactionsByIdOrHashCacheControlHeader(
    isTransactionHash,
    scheduled !== undefined,
    rows
  );
```

the function returns an object with customized headers and it's set to the `responseHeadersLabel` in res.locals. Then `responseHandler` will merge the customized headers (as the last one) from endpoint handler with the default headers.

File: hedera-mirror-rest/schedules.js [hunk 5]
```diff
@@ -134,7 +134,29 @@ const getScheduleById = async (req, res) => {
...
+  }
+  res.locals[constants.responseDataLabel] = formatScheduleRow(schedule);
+};
+
+const calculateScheduleMaxAge = (schedule) => {
+  const ONE_BILLION = BigInt(1_000_000_000);
```
Comment (xin-hedera) [inline hedera-mirror-rest/schedules.js]: the simpler form of a BigInt is `1_000_000_000n`. prefer making this a constant in `constants.js`, similar to `NANOSECONDS_PER_MILLISECOND`.

File: hedera-mirror-rest/schedules.js [hunk 6]
```diff
@@ -134,7 +134,29 @@ const getScheduleById = async (req, res) => {
...
+  const nowNs = utils.nowInNs();
+  const executedTimestamp = schedule.executed_timestamp;
+  const expirationTime = schedule.expiration_time;
+  const consensusTimestamp = schedule.consensus_timestamp;
+
+  const hasExecuted = executedTimestamp !== undefined || schedule.deleted;
+  const hasAutoExpired = expirationTime === undefined && nowNs >= consensusTimestamp +  THIRTY_ONE_MINUTES * ONE_BILLION;
+  const hasExpired = expirationTime !== undefined && nowNs >= expirationTime + SIXTY_SECONDS * ONE_BILLION;
+
+  return hasExecuted || hasAutoExpired || hasExpired ? 3600 : undefined;
```
Comment (xin-hedera) [inline hedera-mirror-rest/schedules.js:159]: make the header a constant in this module, e.g., `const LONGER_SCHEDULE_CACHE_CONTROL_HEADER = {'cache-control': 'public, max-age=3600'}`. Note we use call lower cases.

File: hedera-mirror-rest/schedules.js [hunk 7]
```diff
@@ -134,7 +134,29 @@ const getScheduleById = async (req, res) => {
...
-  res.locals[constants.responseDataLabel] = formatScheduleRow(rows[0]);
+  const schedule = rows[0];
+  const maxAge = calculateScheduleMaxAge(schedule);
+  if (maxAge !== undefined) {
+    res.set('Cache-Control', `public, max-age=${maxAge}`);
+  }
+  res.locals[constants.responseDataLabel] = formatScheduleRow(schedule);
+};
+
+const calculateScheduleMaxAge = (schedule) => {
```
Comment (xin-hedera) [inline hedera-mirror-rest/schedules.js:145]: since the logic is now in its own function, we should add unit tests to `schedules.test.js`

**Review by xin-hedera — CHANGES_REQUESTED** [review]

File: hedera-mirror-rest/schedules.js (hunk 4 above)
Comment (HarshSawarkar) [reply hedera-mirror-rest/schedules.js]: Updated

File: hedera-mirror-rest/schedules.js (hunk 7 above)
Comment (HarshSawarkar) [reply hedera-mirror-rest/schedules.js]: Added the test.

File: hedera-mirror-rest/__tests__/specs/schedules/{id}/executed.json [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+      }
+    ]
+  },
+  "url": "/api/v1/schedules/2000",
+  "responseStatus": 200,
+  "responseHeaders": {
+    "cache-control": "public, max-age=3600"
+  },
+  "responseJson": {
+    "schedules": [
```
Comment (jnels124) [inline hedera-mirror-rest/__tests__/specs/schedules/{id}/executed.json:48]: this shouldn't be a list response

File: hedera-mirror-rest/__tests__/specs/schedules/{id}/explicitly-expired.json [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+  },
+  "responseJson": {
+    "admin_key": {
+      "_type": "ProtobufEncoded",
+      "key": "AQEB"
+    },
+    "consensus_timestamp": "1234567890.000000001",
+    "creator_account_id": "0.0.1024",
+    "deleted": false,
+    "expiration_time": "1234567890.000000001",
```
Comment (jnels124) [inline hedera-mirror-rest/__tests__/specs/schedules/{id}/explicitly-expired.json:39]: should be `1234567890.600000001`

File: hedera-mirror-rest/__tests__/specs/schedules/{id}/explicitly-expired.json [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+  },
+  "url": "/api/v1/schedules/0.0.2001",
+  "responseStatus": 200,
+  "responseHeaders": {
+    "cache-control": "public, max-age=3600"
+  },
+  "responseJson": {
+    "admin_key": {
+      "_type": "ProtobufEncoded",
+      "key": "AQEB"
```
Comment (jnels124) [inline hedera-mirror-rest/__tests__/specs/schedules/{id}/explicitly-expired.json:34]: should be `010101`

File: hedera-mirror-rest/__tests__/specs/schedules/{id}/implicitly-expired.json [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+  },
+  "url": "/api/v1/schedules/0.0.2002",
+  "responseStatus": 200,
+  "responseHeaders": {
+    "cache-control": "public, max-age=3600"
+  },
+  "responseJson": {
+    "admin_key": {
+      "_type": "ProtobufEncoded",
+      "key": "AQEB"
```
Comment (jnels124) [inline hedera-mirror-rest/__tests__/specs/schedules/{id}/implicitly-expired.json:33]: should be `010101`

**Review by jnels124 — CHANGES_REQUESTED** [review]
spec 'not-executed-yet.json' will need to be updated to use fake time. Also can probably delete the no-params.json spec as it doesn't add value with the new test changes.

File: hedera-mirror-rest/__tests__/specs/schedules/{id}/implicitly-expired.json (hunk 11 above)
Comment (HarshSawarkar) [reply hedera-mirror-rest/__tests__/specs/schedules/{id}/implicitly-expired.json]: Updated.

File: hedera-mirror-rest/schedules.js [hunk 12]
```diff
@@ -134,7 +135,26 @@ const getScheduleById = async (req, res) => {
...
-  res.locals[constants.responseDataLabel] = formatScheduleRow(rows[0]);
+  const schedule = rows[0];
+  res.locals[constants.responseHeadersLabel] = getScheduleCacheControlHeader(schedule);
+  res.locals[constants.responseDataLabel] = formatScheduleRow(schedule);
+};
+
+const getScheduleCacheControlHeader = (schedule) => {
+  const nowNs = BigInt(utils.nowInNs());
+  const executedTimestamp = schedule.executed_timestamp;
+  const expirationTime = schedule.expiration_time !== undefined ? BigInt(schedule.expiration_time) : 0n;
```
[the remark this answers is not in the document]
Comment (HarshSawarkar) [reply hedera-mirror-rest/schedules.js]: Updated.

**Comment by jnels124** [comment]
@HarshSawarkar can you address the test failures shown [here](https://github.com/hiero-ledger/hiero-mirror-node/actions/runs/13457490304/job/37651771513?pr=10383)?

File: hedera-mirror-rest/schedules.js [hunk 13]
```diff
@@ -134,7 +135,26 @@ const getScheduleById = async (req, res) => {
...
+
+  const hasExecuted = executedTimestamp || schedule.deleted;
+  const hasAutoExpired =
+    expirationTime === 0n && nowNs >= consensusTimestamp + constants.THIRTY_ONE_MINUTES * constants.ONE_BILLION;
+  const hasExpired = expirationTime !== 0n && nowNs >= expirationTime + constants.SIXTY_SECONDS * constants.ONE_BILLION;
+
+  if (hasExecuted || hasAutoExpired || hasExpired) {
+    return LONGER_SCHEDULE_CACHE_CONTROL_HEADER;
+  }
+  return {};
```
Comment (xin-hedera) [inline hedera-mirror-rest/schedules.js:157]: remove all `BigInt(...)` in the function

- `utils.NowInNs()` already returns bigint
- all timestamp values returned from our sql query for the schedule should be also bigints
- database `NULL` will be javascript `null`, so setting `expirationTime` / `consensusTimestamp` to default value `0n` is unnecessary and make code more complicated, and please make boolean conditions explicit against `null`, e.g., `executedTimestamp !== null`, `expirationTime === null && nowNs >= consensus_timestamp + constants...`

File: hedera-mirror-rest/__tests__/schedules.test.js [hunk 14]
```diff
@@ -189,6 +189,87 @@ const verifyExtractSqlFromScheduleFilters = (filters, expectedQuery, expectedPar
...
 
+describe('getScheduleCacheControlHeader', () => {
+  const LONGER_SCHEDULE_CACHE_CONTROL_HEADER = {'cache-control': 'public, max-age=3600'};
+  const testSpecs = [
+    {
+      name: 'Schedule has executed',
+      input: {
+        executed_timestamp: utils.nowInNs() - 1_000_000_000n,
+        expiration_time: undefined,
+        consensus_timestamp: undefined,
```
Comment (xin-hedera) [inline hedera-mirror-rest/__tests__/schedules.test.js:200]:
```suggestion
        expiration_time: null,
        consensus_timestamp: null,
```

File: hedera-mirror-rest/__tests__/schedules.test.js [hunk 15]
```diff
@@ -189,6 +189,87 @@ const verifyExtractSqlFromScheduleFilters = (filters, expectedQuery, expectedPar
...
+        deleted: false,
+      },
+      expected: LONGER_SCHEDULE_CACHE_CONTROL_HEADER,
+    },
+    {
+      name: 'Schedule is deleted',
+      input: {
+        executed_timestamp: undefined,
+        expiration_time: undefined,
+        consensus_timestamp: undefined,
```
Comment (xin-hedera) [inline hedera-mirror-rest/__tests__/schedules.test.js:210]: consensus_timestamp should never be null / undefined

**Review by xin-hedera — CHANGES_REQUESTED** [review]

File: hedera-mirror-rest/__tests__/schedules.test.js (hunk 15 above)
Comment (HarshSawarkar) [reply hedera-mirror-rest/__tests__/schedules.test.js]: Updated

File: hedera-mirror-rest/__tests__/schedules.test.js [hunk 16]
```diff
@@ -175,6 +175,87 @@ const verifyExtractSqlFromScheduleFilters = (filters, expectedQuery, expectedPar
   expect(limit).toStrictEqual(expectedLimit);
 };
 
+describe('getScheduleCacheControlHeader', () => {
+  const LONGER_SCHEDULE_CACHE_CONTROL_HEADER = {'cache-control': 'public, max-age=3600'};
+  const testSpecs = [
+    {
+      name: 'Schedule has executed',
+      input: {
+        executed_timestamp: '1234567890000000002',
```
Comment (xin-hedera) [inline hedera-mirror-rest/__tests__/schedules.test.js:184]: this is still wrong. it should be a BigInt. some columns can be null so `null` is also appropriate depending on the test case. when it's a string, the arithmetic operations in the function like `+` will again append or prepend a number to the string.
```suggestion
        executed_timestamp: 1234567890000000002n,
```

**Review by xin-hedera — CHANGES_REQUESTED** [review]

File: hedera-mirror-rest/__tests__/schedules.test.js (hunk 16 above)
Comment (HarshSawarkar) [reply hedera-mirror-rest/__tests__/schedules.test.js]: Updated

**Comment by jnels124** [comment]
@HarshSawarkar there are some test failures [here](https://github.com/hiero-ledger/hiero-mirror-node/actions/runs/13559376702/job/37957580308?pr=10383). Can you address these?

**Comment by HarshSawarkar** [comment]
> @HarshSawarkar there are some test failures [here](https://github.com/hiero-ledger/hiero-mirror-node/actions/runs/13559376702/job/37957580308?pr=10383). Can you address these?

Updated.

**Review by jnels124 — DISMISSED** [review]
LGTM

File: hedera-mirror-rest/constants.js [hunk 17]
```diff
@@ -1,6 +1,9 @@
 // SPDX-License-Identifier: Apache-2.0
 
 const NANOSECONDS_PER_MILLISECOND = 10n ** 6n;
+const ONE_BILLION = 1_000_000_000n;
```
Comment (steven-sheehy) [inline hedera-mirror-rest/constants.js:4]:
```suggestion
const NANOS_PER_SECOND = 10n ** 9n;
```

**Review by steven-sheehy — APPROVED** [review]
LGTM just a field could be better named

**Review by xin-hedera — DISMISSED** [review]
LGTM

**Review by jnels124 — APPROVED** [review]

**Review by steven-sheehy — APPROVED** [review]
